### PR TITLE
Timing based shutter position support

### PR DESCRIPTION
**Pull request by svanscho**
Use python timers for feedback, master timers for execution

### CODE REVIEW

File: src/gateway/hal/master_controller_core.py [hunk 1]
```diff
@@ -467,15 +467,17 @@ def load_output_status(self):
...
-        self._master_communicator.do_basic_action(BasicAction(action_type=10,
-                                                              action=1,
-                                                              device_nr=shutter_id))
-
-    def shutter_down(self, shutter_id):
-        self._master_communicator.do_basic_action(BasicAction(action_type=10,
-                                                              action=2,
-                                                              device_nr=shutter_id))
+    def shutter_up(self, shutter_id, timer=None):
+        # TODO: implement timer
```
Comment (svanscho) [inline src/gateway/hal/master_controller_core.py:471]: raise NotImplementedError

File: src/gateway/shutter_controller.py [hunk 2]
```diff
@@ -286,15 +305,43 @@ def _shutter_goto_direction(self, shutter_id, direction, desired_position=None):
...
-        if direction == ShutterEnums.Direction.UP:
-            self._master_controller.shutter_up(shutter_id)
-        elif direction == ShutterEnums.Direction.DOWN:
-            self._master_controller.shutter_down(shutter_id)
-        elif direction == ShutterEnums.Direction.STOP:
+    def _calculate_shutter_timer(self, shutter_id, desired_position, steps=TIME_BASED_SHUTTER_STEPS):
+        ShutterController._validate_position(shutter_id, desired_position, steps)
+        shutter = self._get_shutter(shutter_id)
+        if self._actual_positions[shutter_id] is None:
+            raise RuntimeError('Shutter {0} has unknown actual position'.format(shutter_id))
```
Comment (svanscho) [inline src/gateway/shutter_controller.py:314]: do we catch this?

File: src/gateway/shutter_controller.py [hunk 3]
```diff
@@ -286,15 +305,43 @@ def _shutter_goto_direction(self, shutter_id, direction, desired_position=None):
...
-    def _execute_shutter(self, shutter_id, direction):  # type: (int, str) -> None
-        if direction == ShutterEnums.Direction.UP:
-            self._master_controller.shutter_up(shutter_id)
-        elif direction == ShutterEnums.Direction.DOWN:
-            self._master_controller.shutter_down(shutter_id)
-        elif direction == ShutterEnums.Direction.STOP:
+    def _calculate_shutter_timer(self, shutter_id, desired_position, steps=TIME_BASED_SHUTTER_STEPS):
+        ShutterController._validate_position(shutter_id, desired_position, steps)
+        shutter = self._get_shutter(shutter_id)
+        if self._actual_positions[shutter_id] is None:
```
Comment (svanscho) [inline src/gateway/shutter_controller.py:313]: .get()

File: src/gateway/shutter_controller.py [hunk 4]
```diff
@@ -286,15 +305,43 @@ def _shutter_goto_direction(self, shutter_id, direction, desired_position=None):
...
+            raise RuntimeError('Shutter {0} has unknown actual position'.format(shutter_id))
+        delta_position = desired_position - self._actual_positions[shutter_id]
+        direction = self._get_direction(self._actual_positions[shutter_id], desired_position)
+        if direction == ShutterEnums.Direction.STOP:
+            return 0
+        else:
+            configured_timer = getattr(shutter, 'timer_{0}'.format(direction.lower()))
+            return int(abs(delta_position) / float(steps) * configured_timer)
```
Comment (svanscho) [inline src/gateway/shutter_controller.py:321]: should we limit this to 1 byte?

File: src/gateway/shutter_controller.py [hunk 5]
```diff
@@ -286,15 +305,43 @@ def _shutter_goto_direction(self, shutter_id, direction, desired_position=None):
...
+            elif direction == ShutterEnums.Direction.DOWN:
+                self._master_controller.shutter_down(shutter_id, timer=timer)
+
+    def reset_shutter(self, shutter_id):  # type: (int) -> None
+        # reset shutter to known state
+        shutter = self._get_shutter(shutter_id)
+        configured_timer = getattr(shutter, 'timer_up')
+        start = time.time()
+        self._execute_shutter(shutter_id, ShutterEnums.Direction.UP)
+        while self._actual_positions[shutter_id] != 0:
```
Comment (svanscho) [inline src/gateway/shutter_controller.py:340]: should we queue this?

File: src/gateway/shutter_controller.py [hunk 6]
```diff
@@ -386,14 +439,34 @@ def _report_shutter_state(self, shutter_id, new_state, force_report=False):
...
                         new_state = ShutterEnums.State.STOPPED
+                        abs_position_delta = elapsed_time / float(timer) * self.TIME_BASED_SHUTTER_STEPS
+                        position_delta = -abs_position_delta if direction == ShutterEnums.Direction.UP else abs_position_delta
+                        if shutter.up_down_config:
+                            position_delta = -position_delta
+                        actual_position = self._actual_positions[shutter_id]
+                        if actual_position is not None:
+                            new_actual_position = actual_position + int(position_delta)
+                            self._actual_positions[shutter_id] = ShutterController.clamp_position(shutter, new_actual_position)
+                            self._position_accuracy[shutter_id] = self._position_accuracy.get(shutter_id, 0) - 10
```
Comment (svanscho) [inline src/gateway/shutter_controller.py:461]: move -10 to a CONSTANT

File: src/gateway/shutter_controller.py [hunk 7]
```diff
@@ -106,6 +109,7 @@ def update_config(self, config):  # type: (List[ShutterDTO]) -> None
                     self._actual_positions[shutter_id] = None
                     self._desired_positions[shutter_id] = None
                     self._directions[shutter_id] = ShutterEnums.Direction.STOP
+                    self._position_accuracy[shutter_id] = 100 if shutter_dto.steps else 0
```
Comment (svanscho) [inline src/gateway/shutter_controller.py:112]: move 100 to a constant

File: src/gateway/webservice.py [hunk 8]
```diff
@@ -245,11 +245,11 @@ def _openmotics_api(f, *args, **kwargs):
         status = 503  # Service Unavailable
         data = {'success': False, 'msg': 'maintenance_mode'}
     except CommunicationTimedOutException:
-        logger.error('Communication timeout during API call %s', f.__name__)
+        logger.exception('Communication timeout during API call %s', f.__name__)
```
Comment (svanscho) [inline src/gateway/webservice.py:248]: move to error

File: testing/unittests/gateway_tests/shutter_test.py [hunk 9]
```diff
@@ -214,41 +218,62 @@ def shutter_direction(direction, _shutter_id):
...
-        calls = {}
-        for shutter_id in [0, 1]:
-            message = 'Shutter {0} does not support positioning'.format(shutter_id)
+        #                             +- starting actual position
+        #                             |   +- position to go to
+        #                             |   |   +- expected direction after the call
+        #                             |   |   |                            +- expected BA to be executed
+        calls = {}                  # v   v   v                            v
+        for shutter_id, data in {1: [[0,  50, ShutterEnums.Direction.UP,   'up', 50],  # down = 0, up = 100
+                                     [80, 50, ShutterEnums.Direction.DOWN, 'down', 30]]}.items():
```
Comment (svanscho) [inline testing/unittests/gateway_tests/shutter_test.py:246]: perhaps test this also with a timer that is not set to 100 for timer 0 and timer 1

File: testing/unittests/gateway_tests/shutter_test.py [hunk 10]
```diff
@@ -381,25 +411,32 @@ def validate(_shutter_id, _entry):
...
+                                  1: [0, 99, ShutterEnums.Direction.UP,   (20, ShutterEnums.State.GOING_UP)],  # this shutter is inverted
+                                  2: [0, 79,   ShutterEnums.Direction.DOWN, (20, ShutterEnums.State.GOING_DOWN)]}.items():
             validate(shutter_id, entry)
             self.pubsub._publish_all_events()
 
-        time.sleep(50)  # Standard shutters will be down now
+        time.sleep(50)  # Standard shutters will still be going down
 
         controller._actual_positions[2] = 20  # Simulate position reporting
         master_controller._update_from_master_state({'module_nr': 0, 'status': 0b00010100})  # First shutter motor stop
```
Comment (svanscho) [inline testing/unittests/gateway_tests/shutter_test.py:442]: improve readability

File: testing/unittests/master_tests/master_command_test.py [hunk 11]
```diff
@@ -145,6 +148,20 @@ def test_create_input(self):
         self.assertEqual(21, len(ba_input))
         self.assertEqual(bytearray(b'STRBA\x01\x02\x04' + (b'\x00' * 11) + b'\r\n'), ba_input)
 
+    def test_create_basic_action(self):
```
Comment (svanscho) [inline testing/unittests/master_tests/master_command_test.py:151]: add test for older master versions < 113

File: src/gateway/shutter_controller.py (hunk 7 above)
Comment (svanscho) [reply src/gateway/shutter_controller.py:113]: this is a percentage, not changed.

File: testing/unittests/gateway_tests/shutter_test.py (hunk 9 above)
Comment (svanscho) [reply testing/unittests/gateway_tests/shutter_test.py:246]: Done

**Review by LnL7 — APPROVED** [review]

**Pull request merged**